File: M3Tools/Components/Database/CustomerDatabase.vb
Imports System.ComponentModel
Imports SPPBC.M3Tools.Events
Imports SPPBC.M3Tools.Types

Namespace Database
	Public NotInheritable Class CustomerDatabase
		Private Const path As String = "customers"

		<SettingsBindable(True)>
		<Description("The username to use for the database connection")>
		Public Property Username As String
			Get
				Return dbConnection.Username
			End Get
			Set(value As String)
				dbConnection.Username = value
			End Set
		End Property

		<PasswordPropertyText(True)>
		<SettingsBindable(True)>
		<Description("The password to use for the database connection")>
		Public Property Password As String
			Get
				Return dbConnection.Password
			End Get
			Set(value As String)
				dbConnection.Password = value
			End Set
		End Property

		<Bindable(True)>
		<SettingsBindable(True)>
		<Description("The initial catalog to use for the database connection")>
		Public Property BaseUrl As String
			Get
				Return dbConnection.BaseUrl
			End Get
			Set(value As String)
				dbConnection.BaseUrl = value
			End Set
		End Property

		Public Sub AddCustomer(
				   fName As String, lName As String,
				   Optional addrStreet As String = Nothing, Optional addrCity As String = Nothing,
				   Optional addrState As String = Nothing, Optional addrZip As String = Nothing,
				   Optional phone As String = Nothing, Optional email As String = Nothing)
			AddCustomer(fName, lName, New Address(addrStreet, addrCity, addrState, addrZip), phone, email)
		End Sub


		Public Sub AddCustomer(
				   fName As String, lName As String,
				   Optional address As Address = Nothing,
				   Optional phone As String = Nothing, Optional email As String = Nothing)
			AddCustomer(New Customer(-1, fName, lName, address, phone, email))
		End Sub

		Public Sub AddCustomer(customer As Customer)
			dbConnection.Consume(Method.Post, $"/{path}", JSON.ConvertToJSON(customer))
		End Sub

		Public Function GetCustomers() As CustomersCollection
			Return dbConnection.Consume(Of CustomersCollection)(Method.Get, $"/{path}").Result
		End Function

		Public Function GetCustomer(customerID As Integer) As Customer
			If Not Utils.ValidID(customerID) Then
				Throw New ArgumentException($"Invalid CustomerID provided")
			End If

			Return dbConnection.Consume(Of Customer)(Method.Get, $"/{path}/{customerID}").Result
		End Function

		Public Sub UpdateCustomer(customerID As Integer, fName As String, lName As String, street As String, city As String, state As String, zipCode As String, email As String, phone As String)
			UpdateCustomer(customerID, fName, lName, New Address(street, city, state, zipCode), email, phone)
		End Sub

		Public Sub UpdateCustomer(customerID As Integer, fName As String, lName As String, addr As Address, email As String, phone As String)
			UpdateCustomer(New Customer(customerID, fName, lName, addr, email, phone))
		End Sub

		Private Sub UpdateCustomer(customer As Customer)
			If Not Utils.ValidID(customer.Id) Then
				Throw New ArgumentException($"Invalid CustomerID provided")
			End If

			dbConnection.Consume(Method.Put, $"/{path}/{customer.Id}", JSON.ConvertToJSON(customer))
		End Sub

		Public Sub RemoveCustomer(customerID As Integer)
			If Not Utils.ValidID(customerID) Then
				Throw New ArgumentException($"Invalid CustomerID provided")
			End If

			dbConnection.Consume(Method.Delete, $"/{path}/{customerID}")
		End Sub
	End Class
End Namespace

File: M3Tools/Components/Database/CustomerDatabase.cs
using System;
using System.Threading.Tasks;

using SPPBC.M3Tools.M3API;

namespace SPPBC.M3Tools.Database
{
	/// <summary>
	/// The customer based API calls
	/// </summary>
	public partial class CustomerDatabase
	{
		/// <summary>
		/// Retrieves the specified customer from the database
		/// </summary>
		/// <param name="customerID"></param>
		/// <param name="ct"></param>
		/// <returns></returns>
		/// <exception cref="ArgumentException"></exception>
		public async Task<Types.Customer> GetCustomer(int customerID, System.Threading.CancellationToken ct = default)
			=> Utils.ValidID(customerID)
				? await ExecuteWithResultAsync<Types.Customer>(System.Net.Http.HttpMethod.Get, string.Join(Paths.Separator, Paths.Customers, customerID), string.Empty, ct)
				: throw new ArgumentException($"Invalid CustomerID provided");

		/// <summary>
		/// Retrieves all customers from the database
		/// </summary>
		/// <returns></returns>
		public async Task<Types.CustomerCollection> GetCustomers(System.Threading.CancellationToken ct = default)
			=> await ExecuteWithResultAsync<Types.CustomerCollection>(System.Net.Http.HttpMethod.Get, Paths.Customers, string.Empty, ct);

		/// <summary>
		/// Add a customer to the database
		/// </summary>
		/// <param name="customer"></param>
		/// <param name="ct"></param>
		public async Task<bool> AddCustomer(Types.Customer customer, System.Threading.CancellationToken ct = default)
			=> await ExecuteAsync(System.Net.Http.HttpMethod.Post, Paths.Customers, JSON.ConvertToJSON(customer), ct);

		/// <summary>
		/// Updates the info for a customer in the database
		/// </summary>
		/// <param name="customer"></param>
		/// <param name="ct"></param>
		public async Task<bool> UpdateCustomer(Types.Customer customer, System.Threading.CancellationToken ct = default)
			=> Utils.ValidID(customer.Id)
				? await ExecuteAsync(System.Net.Http.HttpMethod.Put, string.Join(Paths.Separator, Paths.Customers, customer.Id), JSON.ConvertToJSON(customer), ct)
				: throw new ArgumentException($"Invalid CustomerID provided");

		/// <summary>
		/// Removes a customer from the database
		/// </summary>
		/// <param name="customerID"></param>
		/// <param name="ct"></param>
		/// <exception cref="ArgumentException"></exception>
		public async Task<bool> RemoveCustomer(int customerID, System.Threading.CancellationToken ct = default)
			=> Utils.ValidID(customerID)
				? await ExecuteAsync(System.Net.Http.HttpMethod.Delete, string.Join(Paths.Separator, Paths.Customers, customerID), string.Empty, ct)
				: throw new ArgumentException($"Invalid CustomerID provided");
	}
}
